test/libmpv_test: add default track selection testing

Dudemanguy opened the pull request:
The amount of options we have that are related track selection is insane and touching any of this stuff is scary. Add some track selection testing to the CI to hopefully make it slightly less scary to touch any of that. Since trying to work in media files from some outside source would be a pain and need to live in some repo somewhere, it's nicer to just generate dummy files on the fly with ffmpeg during the ci runs and use that. Obviously, the actual tests themselves could be even more thorough (external tracks were not even considered), but this is more than a good enough start for now.

Read this before you submit this pull request:
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Reading this link and following the rules will get your pull request reviewed
and merged faster. Nobody wants lazy pull requests.

Dudemanguy commented:
Just for the record, I excluded the mingw, win32 (laziness) , openbsd (the ffmpeg commands got stuck), and ffmpeg 4.4 (the ffmpeg commands are bugged and give wrong output) from running the track selection tests.

kasper93 commented on ci/generate-test-media.sh:
```diff
@@ -0,0 +1,39 @@
+
+# Create core video, audio, and subtitle streams to construct files out of.
+ffmpeg -y -f lavfi -i testsrc=duration=2:size=1280x720 video.mkv
+ffmpeg -y -f lavfi -i sine=frequency=1000:duration=2 audio.flac
```
Wouldn't be better to do it with mpv?

Dudemanguy replied:
I would personally prefer avoiding mpv's encoding mode.

kasper93 commented on .github/workflows/build.yml:
```diff
@@ -271,6 +271,10 @@ jobs:
 
+      - name: Generate media test files.
+        run: |
+            ./ci/generate-test-media.sh
```
I think it should be called from meson as test prerequisite. After those changes it would be harder to test locally, because you need to generate files.

Or we even create test file repository, same as ffmpeg does with FATE.

Dudemanguy replied:
>Or we even create test file repository

Well this PR is trying to avoid having to do all that of course. Calling it in meson is probably better though yeah.

kasper93 replied:
I would prefer more unit test like testing, instead of black box ones, but I guess it can do.

Dudemanguy replied:
Moved to the meson file now.

kasper93 commented on test/meson.build:
```diff
@@ -117,8 +117,16 @@ test('paths', paths)
+    ffmpeg = find_program('ffmpeg', required: false)
+    if ffmpeg.found() and ffmpeg.version().version_compare('>=0.5')
+        generate_test_media = find_program(join_paths(source_root, 'ci', 'generate-test-media.sh'), required: true)
+        run_command(generate_test_media, build_root, check: true)
```
I would use `custom_target` to generate it on demand and not on every configure.

Dudemanguy replied:
Hmm, `custom_target` is really annoying in this context.

kasper93 replied:
I have not tried myself. But wouldn't it be 

```
test_files = []
test_file += custom_target('eng_default.mkv',
    output: 'eng_default.mkv',
    command: [ffmpeg, ..., '@OUTPUT@'],
)
```
and then 
```
test('libmpv', exe, args: [file, test_files], timeout: 60, should_fail: not features['lua'])
```
remove 
```
const char *files[]
```
array and ready input files from argv.

kasper93 replied:
Generating files on every configure unconditionally with external ffmpeg is no-go for me. Exact solution is up in the air, but I would prefer it not do complex stuff in configure.

Dudemanguy replied:
It would work but you would need a different target for every single generated file which feels kind of stupid. But I haven't thought of anything else better yet.

kasper93 replied:
Well, for me it is natural, they are different commads after all

Dudemanguy replied:
Reworked into custom targets now.

Dudemanguy commented on test/meson.build:
```diff
@@ -117,8 +117,16 @@ test('paths', paths)
+
+    # Generate files for testing if possible. ffmpeg 4 generates incorrect output so skip.
+    ffmpeg = find_program('ffmpeg', required: false)
+    if ffmpeg.found() and ffmpeg.version().version_compare('>=0.5')
```
Actually, the version parsing doesn't even work.

Dudemanguy replied:
Check is no longer relevant. If there's a different ffmpeg version binary in your path for some reason, oh well. Not going to write custom code for that.

sfan5 commented on test/libmpv_test.c:
```diff
@@ -271,9 +288,140 @@ static void test_options_and_properties(void)
         fail("Node: expected 1 but got %d'!\n", result_node.u.flag);
 }
 
+static void test_track_selection(char *directory)
```
how about moving splitting this into a different file and test run?

Dudemanguy replied:
I considered that but this file has a lot of generic useful stuff in it for the test.

sfan5 replied:
I would move them to a shared file.

Dudemanguy replied:
Now it is split.

Dudemanguy commented:
I reworked this so the `libmpv_test.c` file is split into separate files for every single test it does. Relevant functions are shared inline in `libmpv_shared.h` now. `libmpv_test_track_selection.c` uses meson custom targets as well.

Dudemanguy commented:
>openbsd fails the test for some reason. Will take a look later.

Fixed now. It's because openbsd had the old ffmpeg version (4.4) in its path and that version has a bug when setting forced tracks causing the test to fail. I don't believe there is a way for `find_program` to find something at compile time instead of configure time (at least it didn't seem to work for me), so it doesn't seem possible to use the `ffmpeg` binary that would generated during compilation. So instead, I just added some version checks so this particular test is skipped.

kasper93 commented on test/libmpv_test_track_selection.c:
```diff
@@ -0,0 +1,180 @@
+static void test_track_selection(char *directory, char *file)
+{
+    char path[512] = {0};
+    snprintf(path, 512, "%s/%s", directory, file);
```
```suggestion
    snprintf(path, sizeof(path), "%s/%s", directory, file);
```

Dudemanguy commented on test/samples/meson.build:
```diff
@@ -0,0 +1,80 @@
+samples += custom_target('eng_default.mkv',
+    output: 'eng_default.mkv',
+    depends: [video, audio],
+    command: [ffmpeg, '-y', '-i', video.full_path(), '-i', audio.full_path(), '-f', 'srt',
```
If we're using ffmpeg here, it's easier to use the full path instead of constructing whatever the relative path is based on whatever direction meson happens to run this command from.

kasper93 replied:
meson will construct it automatically, based on the context. And commands will look likely more readable.

Dudemanguy replied:
Oh my bad, I didn't even realize it would work simply by passing the target directly.

Dudemanguy commented on test/samples/meson.build:
```diff
@@ -0,0 +1,80 @@
+# Create core video, audio, and subtitle streams to construct files out of.
+video = custom_target('video.mkv',
+    output: 'video.mkv',
+    command: [ffmpeg, '-y', '-f', 'lavfi', '-i', 'testsrc=duration=2:size=1280x720', '@OUTPUT@'],
```
Is it even possible to use mpv's encoding mode to do the metadata tagging per stream? The manual makes it sound like it applies to the whole file which is no good.

kasper93 replied:
Probably not. We can depend on ffmpeg, it's fine.

Dudemanguy commented on test/libmpv_shared.h:
```diff
@@ -0,0 +1,93 @@
+/*
```
Sure.

Dudemanguy commented on test/libmpv_shared.h:
```diff
@@ -0,0 +1,93 @@
+#endif
+
+// Global handle.
+static mpv_handle *ctx;
```
The inline functions need this though.

Dudemanguy commented on test/libmpv_shared.h:
```diff
@@ -0,0 +1,93 @@
+
+        if (ev->event_id == MPV_EVENT_NONE) {
+            continue;
+        } else if (ev->event_id == MPV_EVENT_LOG_MESSAGE) {
```
Not sure why?

kasper93 replied:
I mean 
```
if (ev->event_id == MPV_EVENT_NONE)
    continue;

if (...)
```
it more readable for me, to see early exit.

Dudemanguy replied:
Sure.

Dudemanguy commented on test/libmpv_test_file_loading.c:
```diff
@@ -0,0 +1,64 @@
+{
+    const char *cmd[] = {"loadfile", file, NULL};
+    check_api_error(mpv_command(ctx, cmd));
+    int loaded = 0;
+    int finished = 0;
```
Yeah, I'll just add stdbool to the common header.

Dudemanguy commented on test/samples/meson.build:
```diff
@@ -0,0 +1,80 @@
+samples = []
+
+# Selecting subtitles with only 1 track has different logic so test separately.
+samples += custom_target('eng_default.mkv',
```
I'm not seeing what you're envisioning here.

kasper93 replied:
I was thinking, something like below. Not sure it will look better, but core `custom_target` structure is the same for all of them, except command and name.
```
foo = {
'eng_default.mkv': ['ffmpeg', ...]
'eng_default_forced.mkv':  ['ffmpeg', ...]
}

foreach name, cmd: foo
    samples += custom_target(....)
endforeach
```

Dudemanguy replied:
That's much nicer.

kasper93 commented on test/samples/meson.build:
```diff
@@ -0,0 +1,63 @@
+}
+
+sample_targets = []
+foreach name, cmd: samples
```
Open question. Why not define `test()` for each of them? This will allow to see individual test results without looking up verbose output and remove loop form test itself, so it tests only one unit at the time.

Dudemanguy replied:
That's a good idea. With some minor tweaking it could work.

kasper93 commented on test/samples/meson.build:
```diff
@@ -0,0 +1,63 @@
+endforeach
+
+sample_paths = []
+foreach sample: sample_targets
```
this loop could be merged with above loop now.

kasper93 commented:
> Program ffmpeg found: NO

I think we should ensure that at least on linux ci the ffmpeg is there.

Dudemanguy commented:
>I think we should ensure that at least on linux ci the ffmpeg is there.

It just needs to thrown on the tumbleweed container. cc @llyyr.

kasper93 commented on test/libmpv_test_track_selection.c:
```diff
@@ -0,0 +1,177 @@
+    check_api_error(mpv_get_property(ctx, property, MPV_FORMAT_STRING, &result_string));
+    if (strcmp(expect, result_string) != 0)
+        fail("String: expected '%s' but got '%s'!\n", expect, result_string);
+    mpv_free(result_string);
```
This should be freed also in fail case. To avoid bogus leak detector warnings on errors.

Dudemanguy replied:
ASAN doesn't actually complain about this. I can make it fail on purpose and there's no leak message.

kasper93 replied:
Ok, fair, I guess. It doesn't report, because by the time we exit the pointer is still reachable on stack, so it's not considered leak.

kasper93 commented on test/samples/meson.build:
```diff
@@ -0,0 +1,62 @@
+
+samples = {
+    'eng_default.mkv':
+        [ffmpeg, '-y', '-i', video, '-i', audio, '-f', 'srt', '-i', sub,
```
Please add `'-v', 'error'` or something similar to not spam logs with those commands.

kasper93 commented on test/samples/meson.build:
```diff
@@ -0,0 +1,62 @@
+samples = {
+    'eng_default.mkv':
+        [ffmpeg, '-y', '-i', video, '-i', audio, '-f', 'srt', '-i', sub,
+         '-map', '0:0', '-map', '1:0', '-map', '2:0', '-c:v', 'copy',
```
Do you need those map commands? I think it would merge all into one by default?

Dudemanguy replied:
Ah yeah it's superfluous for all of them except the multilang one.

kasper93 commented on test/samples/meson.build:
```diff
@@ -0,0 +1,59 @@
+
+samples = {
+    'eng_default.mkv':
+        [ffmpeg, '-v', 'error', '-y', '-i', video, '-i', audio, '-f', 'srt', '-i', sub,
```
We could also move common part of this command to separate variable to make it more readable. Also `'-c:v', 'copy', '-c:a', 'copy', '-c:s', 'srt'` likely can be replaced with simpler `'-c', 'copy', '-c:s', 'srt'`

Dudemanguy replied:
Simplified

kasper93 reviewed (approved):
Looks ok.

llyyr commented:
> I think we should ensure that at least on linux ci the ffmpeg is there.

Done, the new image should be there in 15-20 mins.

Dudemanguy commented:
>Program ffmpeg found: YES 7.1 7.1 (/usr/bin/ffmpeg)

Thanks, it's there now.

kasper93 commented on test/libmpv_test_track_selection.c:
```diff
@@ -0,0 +1,177 @@
+        check_api_error(mpv_set_property_string(ctx, "subs-fallback", "default"));
+    } else if (strcmp(file, "eng_default_forced.mkv") == 0) {
+        // --subs-fallback-forced=no
+        check_api_error(mpv_set_property_string(ctx, "subs-fallback-forced", "no"));
```
Could you add also a test for `subs-fallback-forced=always`? On #mpv we got report that `subs-fallback=no` didn't select default track as expected. But `subs-fallback-forced=always` did select non-forced default track. Which sounds broken, reading the docs it shouldn't fallback to non-forced track.

Dudemanguy replied:
I did but it still passed.

kasper93 replied:
I may have misunderstood the issue. But tracking it back, I think `subs-with-matching-audio=forced` has some issue. See this dirty test I added quickly, it select default track, while as I understand it should select forced, no?

diff --git a/test/libmpv_test_track_selection.c b/test/libmpv_test_track_selection.c
index edd9623d8d..fb30c1603f 100644
--- a/test/libmpv_test_track_selection.c
+++ b/test/libmpv_test_track_selection.c
@@ -46,7 +46,15 @@ static void test_track_selection(char *file, char *path)
         fail("Test file, '%s', was not found!\n", path);
 
     const char *cmd[] = {"loadfile", path, NULL};
-    if (strcmp(file, "eng_default.mkv") == 0) {
+    if (strcmp(file, "foo.mkv") == 0) {
+        check_api_error(mpv_set_property_string(ctx, "slang", "eng"));
+        check_api_error(mpv_set_property_string(ctx, "subs-with-matching-audio", "forced"));
+        check_api_error(mpv_command(ctx, cmd));
+        wait_for_file_load();
+        check_string("current-tracks/sub/selected", "yes");
+        check_string("current-tracks/sub/default", "no");
+        check_string("current-tracks/sub/forced", "yes");
+    } else if (strcmp(file, "eng_default.mkv") == 0) {
         // --no-config
         check_api_error(mpv_command(ctx, cmd));
         wait_for_file_load();
diff --git a/test/samples/meson.build b/test/samples/meson.build
index 3236d9d3f2..d427015de2 100644
--- a/test/samples/meson.build
+++ b/test/samples/meson.build
@@ -18,6 +18,12 @@ common_args = ['-v', 'error', '-y', '-i', video, '-i', audio,
               '-metadata:s:s:0', 'language=eng']
 
 samples = {
+    'foo.mkv':
+        [ffmpeg, '-v', 'error', '-y', '-i', video, '-i', audio,
+               '-f', 'srt', '-i', sub, '-i', sub, '-map', '0', '-map', '1', '-map', '2', '-map', '3', '-c', 'copy',
+              '-metadata:s:a:1', 'language=eng',
+              '-metadata:s:s:0', 'language=eng', '-disposition:s:0', 'default',
+              '-metadata:s:s:1', 'language=eng', '-disposition:s:1', 'forced', '@OUTPUT@'],
     'eng_default.mkv':
         [ffmpeg, common_args, '-disposition:s:0', 'default', '@OUTPUT@'],
     'eng_default_forced.mkv':

kasper93 commented on test/samples/meson.build:
```diff
@@ -0,0 +1,57 @@
+         '-disposition:s:0', 'forced', '@OUTPUT@'],
+    'eng_no_default.mkv':
+        [ffmpeg, common_args, '@OUTPUT@'],
+    'multilang.mkv':
```
`-c copy` is missing here.

sfan5 commented on test/libmpv_test_track_selection.c:
```diff
@@ -0,0 +1,193 @@
+{
+    FILE *fp = fopen(path, "r");
+    if (!fp)
+        fail("Test file, '%s', was not found!\n", path);
```
missing fclose.
consider [`access(3)`](https://man.archlinux.org/man/access.3p) (unless mingw somehow doesn't have that)

Dudemanguy replied:
Changed to `access`

sfan5 reviewed (approved):
looks good otherwise
